src/sync/impl/apple/network_reachability_observer.cpp

```diff
@@ -0,0 +1,131 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync/impl/apple/network_reachability_observer.hpp"
+
+#if !TARGET_OS_WATCH
+
+using namespace realm;
+
+namespace {
+
+static NetworkReachabilityStatus reachability_status_for_flags(SCNetworkReachabilityFlags flags)
+{
+    if (!(flags & kSCNetworkReachabilityFlagsReachable)) {
+        return NotReachable;
+    }
+
+    NetworkReachabilityStatus status = NotReachable;
+
+    if (!(flags & kSCNetworkReachabilityFlagsConnectionRequired)) {
+        status = ReachableViaWiFi;
+    }
+
+    if ((flags & kSCNetworkReachabilityFlagsConnectionOnTraffic) &&
+        !(flags & kSCNetworkReachabilityFlagsInterventionRequired)) {
+        status = ReachableViaWiFi;
+    }
+
+#if TARGET_OS_IPHONE
+    if (flags & kSCNetworkReachabilityFlagsIsWWAN) {
+        status = ReachableViaWWAN;
+    }
+#endif
+
+    return status;
+}
+
+static void reachability_callback(SCNetworkReachabilityRef, SCNetworkReachabilityFlags, void* info)
+{
+    auto callback = reinterpret_cast<std::function<void ()> *>(info);
+    (*callback)();
+}
+
+} // (anonymous namespace)
+
+NetworkReachabilityObserver::NetworkReachabilityObserver(util::Optional<std::string> hostname,
+                                                         std::function<void (const NetworkReachabilityStatus)> handler)
+: m_callback_queue(dispatch_queue_create("io.realm.sync.reachability", DISPATCH_QUEUE_SERIAL))
+, m_reachability_callback([=]() { reachability_changed(); })
+, m_change_handler(std::move(handler))
+{
+    if (hostname) {
+        m_reachability_ref = util::adoptCF(SCNetworkReachabilityCreateWithName(nullptr, hostname->c_str()));
+    } else {
+        struct sockaddr zeroAddress = {};
+        zeroAddress.sa_len = sizeof(zeroAddress);
+        zeroAddress.sa_family = AF_INET;
+
+        m_reachability_ref = util::adoptCF(SCNetworkReachabilityCreateWithAddress(nullptr, &zeroAddress));
+    }
+}
+
+NetworkReachabilityObserver::~NetworkReachabilityObserver()
+{
+    stop_observing();
+    dispatch_release(m_callback_queue);
+}
+
+NetworkReachabilityStatus NetworkReachabilityObserver::reachability_status() const
+{
+    SCNetworkReachabilityFlags flags;
+
+    if (SCNetworkReachabilityGetFlags(m_reachability_ref.get(), &flags)) {
+        return reachability_status_for_flags(flags);
+    }
+
+    return NotReachable;
+}
+
+bool NetworkReachabilityObserver::start_observing()
+{
+    m_previous_status = reachability_status();
+
+    SCNetworkReachabilityContext context = {0, &m_reachability_callback, nullptr, nullptr, nullptr};
```

Comment:
The `std::function` is not needed for that. This can just be:

    auto callback = [](void* self) { static_cast<NetworkReachabilityObserver*>(self)->reachability_changed(); };
    SCNetworkReachabilityContext context = {0, this, nullptr, nullptr, nullptr};
    if (!SystemConfiguration::shared().network_reachability_set_callback(m_reachability_ref.get(), callback, &context)) {